audbackend/core/artifactory.py

```diff
@@ -1,13 +1,159 @@
 
+import artifactory
 import dohq_artifactory
 
-import audfactory
+import audeer
 
 from audbackend.core import utils
 from audbackend.core.backend import Backend
 
 
+def _artifactory_path(
+        path,
+        username,
+        apikey,
+) -> artifactory.ArtifactoryPath:
+    r"""Authenticate at Artifactory and get path object."""
+    return artifactory.ArtifactoryPath(
+        path,
+        auth=(username, apikey),
+    )
+
+
+def _authentication(host) -> typing.Tuple[str, str]:
+    """Look for username and API key.
+
+    It first looks for the two environment variables
+    ``ARTIFACTORY_USERNAME`` and
+    ``ARTIFACTORY_API_KEY``.
+
+    If some of them or both are missing,
+    it tries to extract them from the
+    :file:`~/.artifactory_python.cfg` config file.
+    For that it removes ``http://`` or ``https://``
+    from the beginning of ``url``
+    and ``/`` from the end.
+    E.g. for ``https://audeering.jfrog.io/artifactory/`
+    it will look for an entry in the config file under
+    ``[audeering.jfrog.io/artifactory]``.
+
+    If it cannot find the config file
+    or a matching entry in the config file
+    it will set the username to ``anonymous``
+    and the API key to an empty string.
+    If your Artifactory server is configured
+    to allow anonymous users
+    you will be able to access the server this way.
+
+    Args:
+        host: host address
+
+    Returns:
+        username and API key
+
+    """
+    username = os.getenv('ARTIFACTORY_USERNAME', None)
+    apikey = os.getenv('ARTIFACTORY_API_KEY', None)
+    if apikey is None or username is None:  # pragma: no cover
+        if host.startswith('http://'):
+            host = host[7:]
+        elif host.startswith('https://'):
+            host = host[8:]
+        if host.endswith('/'):
+            host = host[:-1]
+        config_entry = artifactory.get_global_config_entry(host)
+        if config_entry is None:
+            if username is None:
+                username = 'anonymous'
+            if apikey is None:
+                apikey = ''
+        else:
+            if username is None:
+                username = config_entry['username']
+            if apikey is None:
+                apikey = config_entry['password']
+    return username, apikey
+
+
+def _deploy(
+        src_path: str,
+        dst_path: artifactory.ArtifactoryPath,
+        *,
+        verbose: bool = False,
+):
+    r"""Deploy local file as an artifact.
+
+    Args:
+        src_path: local file path
+        dst_path: path on Artifactory
+        verbose: show information on the upload process
+
+    """
+    if verbose:  # pragma: no cover
+        desc = audeer.format_display_message(
+            f'Deploy {src_path}',
+            pbar=False,
+        )
+        print(desc, end='\r')
+
+    md5 = utils.md5(src_path)
+
+    if not dst_path.parent.exists():
+        dst_path.parent.mkdir()
+    with open(src_path, 'rb') as fd:
+        dst_path.deploy(fd, md5=md5)
```

Comment:
In the function from `audfactory` also `sha1` and `sha256` were calculated and passed to the function. But in `audbackend` we only rely on `md5`, so I am not sure if we want this overhead?

Reply:
And there is a `explode_archive_atomic` argument to the `deploy()` function:

![image](https://user-images.githubusercontent.com/10383417/234695603-da76e573-7008-46af-a1db-b4da45ae6c84.png)

I wonder if this would ensures an artifact is only uploaded when the operation was successful (see #108). If yes and if it does not have any negative effects, we should probably turn it on. This would avoid a user has to ask an administrator to delete corrupted files if she's lacking permission to delete artifacts.

Reply:
But doesn't `explode_archive_atomic` only apply when `explode_archive` is `True`?
And if I understand it correctly storing an archive exploded means it is stored as a folder containing all the files and not as an archive, compare https://stackoverflow.com/a/6980228.

Reply:
Ok, too bad they don't support deploying a regular file as an atomic operation.

Reply:
Regarding the checksums, you are most likely right. When I interpret https://jfrog.com/help/r/jfrog-rest-apis/deploy-artifact-by-checksum correctly there is no need at all to provide any checksums when deploying. So I guess it will be fine to just pass on the MD5 sum.